File: Data_Structure/Queue/Circular_Queue.c
#include<stdio.h>
#include<stdlib.h>
#define MAX 5

int queueArray[5];
int front = -1;
int rear = -1;

//Check if the queue is full
int isFull()
{
    if((front == rear + 1) || (front == 0 && rear == MAX-1))
    {
        return 1;
    }
    return 0;
}

//Check if the queue is empty
int isEmpty()
{
    if(front == -1)
    {
        return 1;
    }
    return 0;
}

//Adding an element
void enQueue(int element)
{
    if(isFull())
    {
        printf("Stack is full !!\n\n");
        return;
    }
    else
    {
        if(front == -1)
        {
            front = 0;  
        }
        rear = (rear+1)%MAX;
        queueArray[rear] = element;
    }
}

//Removing an element
void deQueue()
{
    int x;
    if(isEmpty())
    {
        printf("Stack is empty !!\n\n");
        return;
    }
    else
    {
        x = queueArray[front];
        if(front == rear)
        {
            front = rear = -1;  
        }
        else
        {
            front = (front+1)%MAX;
        }
        printf("\nDeleted element is: %d", queueArray[front]);
    }
}

//Size of queue
int size()
{
    int i, n = 0;
    for(i = 0; i<=MAX-1; i++)
    {
        if(queueArray[i] != 0)
        {
            n += 1;
        }
    }
    printf("\nSize of queue is %d",n);
    return n;
}

//Display the queue
void display()
{
    int i;
    if(front == -1)
    {
        printf("Queue is empty");
    }
    else
    {
        for(i=front; i!=rear; i=(i+1)%MAX)
        {
            printf("%d ",queueArray[i]);
        }
        printf("%d ", queueArray[i]);
        printf("\n");
    }
}

int main()
{
    enQueue(45);
    enQueue(12);
    enQueue(33);
    enQueue(41);
    enQueue(54);
    //enQueue(23);
    printf("Queue of elements is:\n");
    display();
    
    deQueue();
    printf("\n\nDequeue after deleting element at rear is:\n");
    display();
    
    size();
    return 0;
}

File: Data_Structure/DS_Using_Python/Circular_Queue.py
class Queue:
    def __init__(self, capacity):
        self.capacity = capacity
        self.queue = [None]*capacity
        self.head = 0
        self.tail = -1
        self.size = 0

    # Add an element
    def enqueue(self, item):
        if self.size == self.capacity:
        	print("Stack is full !!")
        else:
        	self.tail = (self.tail+1)%self.capacity
        	self.queue[self.tail] = item
        	self.size = self.size + 1

    # Remove an element
    def dequeue(self):
        if self.size == 0:
            print("Stack is empty !!")
            return
        else:
        	x = self.queue[self.head]
        	self.head = (self.head+1)%self.capacity
        self.size = self.size - 1
        print("\n\nDelete item is {}".format(x))

    # Display  the queue
    def display(self):
        if self.size == 0:
        	print("Stack is empty !!")
        else:
        	index = self.head
        	for i in range(self.size):
        		print(self.queue[index], end=' ')
        		index = (index+1)%self.capacity

	#Size of queue
    def Queuesize(self):
        return self.size
        
    #find an element in queue
    def search(self, val):
    	for i in range(len(self.queue)):
    		   if self.queue[i] == val:
    		      print("\n\nElement found at {} position.".format(i))
    		      return
    	print("\n\nNo element found")
    	
    #top most element in queue
    def peek(self):
    	print("\nPeek of queue is:", self.queue[0])

q = Queue(5)
q.enqueue(1)
q.enqueue(2)
q.enqueue(3)
q.enqueue(34)
q.enqueue(5)

print("Queue of elements is:")
q.display()

q.dequeue()
print("\nAfter removing an element:")
q.display()

q.search(34)
q.peek()

sizeOfQueue = q.Queuesize()
print('\nSize of queue is', sizeOfQueue)